File: python3/merge_k_sorted_lists.py
"""
Space   : O(n)
Time    : O(n log n)
"""


class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        n = 0
        ans = ListNode(0)
        p = ans
        temp = []

        for x in lists:
            while x:
                temp.append(x.val)
                x = x.next
                n += 1

        temp.sort()

        for i in range(n):
            p.next = ListNode(temp[i])
            p = p.next

        return ans.next

File: golang/merge_k_sorted_lists.go
package main

// Space 	: O(nk)
// Time 	: O(nk)

import (
	"golang/model"
	"sort"
)

func mergeKLists(lists []*model.ListNode) *model.ListNode {
	temp := []int{}
	ans := &model.ListNode{Val: 0, Next: nil}
	k := len(lists)
	p := &model.ListNode{}

	for i := 0; i < k; i++ {
		p = lists[i]
		for p != nil {
			temp = append(temp, p.Val)
			p = p.Next
		}
	}

	sort.Ints(temp)
	p = ans
	for _, val := range temp {
		p.Next = &model.ListNode{Val: val, Next: nil}
		p = p.Next
	}

	return ans.Next
}
